### ISSUE TITLE
accessToken not being renewed?

### ISSUE THREAD

**Issue by cchew** [ForgeRock/appAuthHelper#16]
Hi there,

We are noticing that our stale accessToken is not being renewed when calling resourceServer.

This is the response from the service call:

```
{  
   "error_description":"The access token provided is expired, revoked, malformed, or invalid for other reasons.",
   "error":"invalid_token"
}
```

We can see 2 calls made:

![image](https://user-images.githubusercontent.com/13617899/68717216-453fe680-05fa-11ea-8284-9a9ee7f46948.png)

Interestingly the first GET call doesn't include the 'Authorization' header (same response as above) while the second GET call includes the 'Authorization' header with the stale accessToken.

Is there anything else that we need to do?

Thank you!

**Comment by jakefeasel** [comment]
AppAuthHelper looks for the "invalid_token" error as part of the WWW-Authenticate header, as described here: https://tools.ietf.org/html/rfc6750#section-3.1 . As it says there, this is the case where "The client MAYrequest a new access token and retry the protected resource request."

The example provided is as follows:
```
WWW-Authenticate: Bearer realm="example",
                       error="invalid_token",
                       error_description="The access token expired"
```

Is your RS including this response header?

**Comment by cchew** [comment]
Ah, our RP (/userinfo) is OpenAM 13 which is returning the error+description in the response body. Will see if we can configure it to return the header that you outlined.

**Comment by jakefeasel** [comment]
Yes, this is actually an open bug in AM, you can see it described here: https://bugster.forgerock.org/jira/browse/OPENAM-11493 

You probably won't be able to configure AM to return the error differently, but there are other options - 
1) If you have a reverse proxy in front of AM (such as IG) then you could intercept this response and add the appropriate header. 
2) You could modify AppAuthHelper to look for the error in the response body in addition to the header.

**Comment by fetimo** [comment]
I'm also experiencing this issue but the header is set. In the Network tab the request has two entries, one for the service worker and one for the main thread. I can see that the server responds with the following headers on the service worker with:

```
www-authenticate: Bearer, error="invalid_token", error_description="The access token expired"
```

And on the main thread it displays as:

```
www-authenticate: Bearer
www-authenticate: error="invalid_token"
www-authenticate: error_description="The access token expired"
```

I'm not sure if this difference is a quirk of the dev tool or if the service worker splits them out into separate headers. Does the casing of the header matter?

**Comment by jakefeasel** [comment]
The case doesn't matter - http headers are case-insensitive by design. However, I can see how the response you're getting is incompatible with the current code. Here's the current logic:

```javascript
        getAuthHeaderDetails: function (resp) {
            var authHeader = resp.headers.get("www-authenticate");

            if (!resp.ok && authHeader && authHeader.match(/^Bearer /)) {
                return authHeader.replace(/^Bearer /, "")
                    .match(/[^,=]+=".*?"/g)
                    .reduce(function (result, detail) {
                        var pair = detail.split("=");
                        result[pair[0]] = pair[1].replace(/"(.*)"/, "$1");
                        return result;
                    }, {});
            } else {
                return {};
            }
        },
```

I think what's happening here is that your RS is returning multiple `www-authenticate` headers. `headers.get` will return a single string, with each value separated by commas (ref: https://developer.mozilla.org/en-US/docs/Web/API/Headers/get ).

As you can see from the above logic, my regex doesn't handle this csv form of the error. Looks like we need to account for it.

**Comment by jakefeasel** [comment]
@fetimo could you try out that PR and see if it works for you? Thanks

**Comment by jakefeasel** [comment]
Great, thanks for your help @fetimo . This has been merged and is now available as v 0.3.18. Will be published on NPM once they have it working again - https://status.npmjs.org/ reports issues with publishing.
